distutils/extension.py

```diff
@@ -106,12 +108,23 @@ def __init__(
     ):
         if not isinstance(name, str):
             raise AssertionError("'name' must be a string")  # noqa: TRY004
-        if not (
-            isinstance(sources, list)
-            and all(isinstance(v, (str, os.PathLike)) for v in sources)
-        ):
+
+        # we handle the string case first; though strings are iterable, we disallow them
+        if isinstance(sources, str):
+            raise AssertionError(  # noqa: TRY004
+                "'sources' must be an iterable of strings or PathLike objects, not a string"
+            )
+
+        # mow we check if it's iterable and contains valid types
+        try:
+            sources = list(sources)  # convert to list for consistency
+            if not all(isinstance(v, (str, os.PathLike)) for v in sources):
+                raise AssertionError(
+                    "All elements in 'sources' must be strings or PathLike objects"
+                )
```

Comment:
On one hand this prevents issues with single-use iterators, one the other it seems like a lot of iterations over the same source over and over.

`os.fspath` already does type-checking, this entire check just seems useless to me

```suggestion
            self.sources = list(map(os.fspath, sources))
```
(and remove the `self.sources = list(map(os.fspath, sources))` below)

Reply:
> `os.fspath` already does type-checking

Interesting, didn't know that! Addressed in 4a467fac7a98cc5d605afc1ebd951cdd82268f86.